### DISCUSSION TITLE
unable to wakeup from Stop2 mode

### DISCUSSION THREAD

**Discussion by RahulVilasPatil** [zephyrproject-rtos/zephyr#53439] [Q&A]
I have done below setting for stop2 mode and try to set PC13 as wakeup source. I am unable to find out what is wrong with below code, please guide me how I can use PC13 as wakeup source.  

RCC_OscInitTypeDef RCC_OscInitStruct = {0};
RCC_ClkInitTypeDef RCC_ClkInitStruct = {0};
	
  if (HAL_PWREx_ControlVoltageScaling(PWR_REGULATOR_VOLTAGE_SCALE1) != HAL_OK)
  {
     Error_Handler();
  }

  RCC_OscInitStruct.OscillatorType = RCC_OSCILLATORTYPE_HSI|RCC_OSCILLATORTYPE_MSI;
  RCC_OscInitStruct.HSIState = RCC_HSI_ON;
  RCC_OscInitStruct.HSICalibrationValue = RCC_HSICALIBRATION_DEFAULT;
  RCC_OscInitStruct.MSIState = RCC_MSI_ON;
  RCC_OscInitStruct.MSICalibrationValue = RCC_MSICALIBRATION_DEFAULT;
  RCC_OscInitStruct.MSIClockRange = RCC_MSIRANGE_4;
  RCC_OscInitStruct.PLL.PLLState = RCC_PLL_ON;
  RCC_OscInitStruct.PLL.PLLSource = RCC_PLLSOURCE_MSI;
  RCC_OscInitStruct.PLL.PLLMBOOST = RCC_PLLMBOOST_DIV1;
  RCC_OscInitStruct.PLL.PLLM = 1;
  RCC_OscInitStruct.PLL.PLLN = 80;
  RCC_OscInitStruct.PLL.PLLP = 1;
  RCC_OscInitStruct.PLL.PLLQ = 1;
  RCC_OscInitStruct.PLL.PLLR = 1;
  RCC_OscInitStruct.PLL.PLLRGE = RCC_PLLVCIRANGE_0;
  RCC_OscInitStruct.PLL.PLLFRACN = 0;

  if (HAL_RCC_OscConfig(&RCC_OscInitStruct) != HAL_OK)
  {
    Error_Handler();
  }

  RCC_ClkInitStruct.ClockType = RCC_CLOCKTYPE_HCLK|RCC_CLOCKTYPE_SYSCLK
                              |RCC_CLOCKTYPE_PCLK1|RCC_CLOCKTYPE_PCLK2
                              |RCC_CLOCKTYPE_PCLK3;
  RCC_ClkInitStruct.SYSCLKSource = RCC_SYSCLKSOURCE_PLLCLK;
  RCC_ClkInitStruct.AHBCLKDivider = RCC_SYSCLK_DIV1;
  RCC_ClkInitStruct.APB1CLKDivider = RCC_HCLK_DIV1;
  RCC_ClkInitStruct.APB2CLKDivider = RCC_HCLK_DIV1;
  RCC_ClkInitStruct.APB3CLKDivider = RCC_HCLK_DIV1;

  if (HAL_RCC_ClockConfig(&RCC_ClkInitStruct, FLASH_LATENCY_4) != HAL_OK)
  {
	  	  Error_Handler();
  }

setting power and peripheral clocks

        LL_AHB3_GRP1_EnableClock(LL_AHB3_GRP1_PERIPH_PWR);
	LL_AHB2_GRP1_EnableClock(LL_AHB2_GRP1_PERIPH_ALL);
	LL_AHB2_GRP1_EnableClockStopSleep(LL_AHB2_GRP1_PERIPH_ALL);
clearing stop flag
     LL_PWR_ClearFlag_STOP();

LL_EXTI_InitTypeDef conf = {0};
EXTI_HandleTypeDef hand = {0};
EXTI_ConfigTypeDef extconf = {0};
	
	__enable_irq();

	hand.RisingCallback = myRisingCallback;
	hand.FallingCallback = myFallingCallback;
	hand.Line = EXTI_LINE_13;
	
	extconf.Line = EXTI_LINE_13;
	extconf.Mode = EXTI_MODE_INTERRUPT;
	extconf.Trigger = EXTI_TRIGGER_RISING_FALLING;
	extconf.GPIOSel = EXTI_GPIOC;
	
                conf.Line_0_31 = LL_EXTI_LINE_9;
		conf.LineCommand = ENABLE;
		conf.Mode = LL_EXTI_MODE_IT_EVENT;
		conf.Trigger = LL_EXTI_TRIGGER_RISING_FALLING;

	HAL_EXTI_ClearPending(&hand,EXTI_TRIGGER_RISING);

#if STM32_SYSCLK_SRC_MSI
#define RCC_STOP_WAKEUPCLOCK_SELECTED LL_RCC_STOP_WAKEUPCLOCK_MSI
#else
#define RCC_STOP_WAKEUPCLOCK_SELECTED LL_RCC_STOP_WAKEUPCLOCK_HSI
#endif
	LL_RCC_SetClkAfterWakeFromStop(RCC_STOP_WAKEUPCLOCK_SELECTED);

		
		 
		LL_EXTI_Init(&conf); 
		HAL_EXTI_SetConfigLine(&hand,&extconf);
		HAL_NVIC_ClearPendingIRQ(EXTI13_IRQn);
		HAL_NVIC_EnableIRQ(EXTI13_IRQn);
		
		LL_LPM_EnableDeepSleep();
		LL_LPM_EnableSleepOnExit();
		LL_LPM_EnableEventOnPend();
		LL_PWR_SetPowerMode(LL_PWR_STOP2_MODE);
		LL_PWR_EnablePUPDConfig();
		LL_PWR_EnableGPIOPullDown(LL_PWR_GPIO_PORTC, LL_PWR_GPIO_PIN_13);
		LL_PWR_SetWakeUpPinPolarityHigh(LL_PWR_WAKEUP_PIN2);
		LL_PWR_EnableWakeUpPin(LL_PWR_WAKEUP_PIN2);
		LL_PWR_SetWakeUpPinSignal1Selection(LL_PWR_WAKEUP_PIN2);
		LL_PWR_ClearFlag_WU();
	
__disable_irq();
    __WFI();
HAL_ResumeTick();

Please guide me I don't know what I have missing here

**Comment by erwango** [comment]
Moving to discussion as this is not a bug but a question.

You are not supposed to use all this code.
All you need to do is to use the wake up pin code:
```
	LL_PWR_SetWakeUpPinPolarityHigh(LL_PWR_WAKEUP_PIN2);
	LL_PWR_EnableWakeUpPin(LL_PWR_WAKEUP_PIN2);
	LL_PWR_SetWakeUpPinSignal1Selection(LL_PWR_WAKEUP_PIN2);
```
All the clock and power parts are already handled by Zephyr driver (assuming your target is correctly configured).

**Reply by RahulVilasPatil** [reply]
how I can add wakeup source in zephyr. I had try to add wakeup source in dts file but it not work I am not able to understand how I can add wakeup source in my code

**Reply by erwango** [reply]
It's not possible to add pin wake up source in device tree yet. You need to use the 3 lines I mentionned

**Reply by rahulvilaspatil-eaton** [reply]
Thanks..., I have use stop2 mode. whatever you have mentioned here is worked with standby perfectly but negative for stop2. yesterday I had gone through some workshop for u5 and there they mention stop2 mode is run with SRD(Smart run mode). I did required setting but still not work.

**Reply by erwango** [reply]
What is you board configuration (dts)?

**Reply by rahulvilaspatil-eaton** [reply]
/ {

	chosen {
		zephyr,console = &usart1;
		zephyr,shell-uart = &usart1;
	};
	
	power-states {
		runtime_idle: runtime-idle {
			compatible = "zephyr,power-state";
			power-state-name = "runtime-idle";
			substate-id = <1>;
			min-residency-us = <110000>;
			exit-latency-us = <100>;
		};
		suspend_to_idle: suspend-to-idle {
			compatible = "zephyr,power-state";
			power-state-name = "suspend-to-idle";
			substate-id = <2>;
			min-residency-us = <120000>;
			exit-latency-us = <100>;
		};
		standby: standby {
			compatible = "zephyr,power-state";
			substate-id = <3>;
			power-state-name = "standby";
			min-residency-us = <130000>;
			exit-latency-us = <100>;
		};
	};
    aliases {
        
        sw1 = &button1;
    };
    gpio_keys {
        compatible = "gpio-keys";
        button1: button_1 {
            gpios = <&gpioc 13 ( GPIO_ACTIVE_HIGH | GPIO_PULL_DOWN)>;
            label = "Button 1";
        };
    };
    
};

&cpu0 {
	cpu-power-states = <&runtime_idle &suspend_to_idle &standby>;
};

&clk_lse {
	status = "okay";
};
&clk_lsi {
	status = "okay";
};

&clk_hsi {
	/* Comment out this line to use HSI as clk source */
	status = "okay";
};
&clk_msis {
	status = "okay";
	msi-range = <4>;
	msi-pll-mode;
};


&pll1 {
	div-m = <1>;
	mul-n = <80>;
	div-q = <4>;
	div-r = <4>;
	clocks = <&clk_msis>;
	status = "okay";
};

&rcc {
	clocks = <&pll1>;
	clock-frequency = <DT_FREQ_M(80)>;
	ahb-prescaler = <1>;
	apb1-prescaler = <1>;
	apb2-prescaler = <1>;
	apb3-prescaler = <1>;
};

**Reply by rahulvilaspatil-eaton** [reply]
this is dts file for stm32u575zi_q

**Reply by erwango** [reply]
`stop2` is not configured in your dts. Have a look to :
https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/b_u585i_iot02a/b_u585i_iot02a-common.dtsi
